docs: New getting started page (#542)
apify/apify-client-js: docs/index.md

---
sidebar_label: 'Getting started'
title: 'Getting started'
---

# Apify API client for JavaScript

> `apify-client` is the official library to access [Apify REST API](https://docs.apify.com/api/v2) from your

Reviewer: ```suggestion
`apify-client` is the official library to access the [Apify REST API](https://docs.apify.com/api/v2) from your
```

JavaScript/TypeScript applications. It runs both in Node.js and browser and provides useful features like
automatic retries and convenience functions that improve the experience of using the Apify API. All requests and responses (including errors) are encoded in JSON format with UTF-8 encoding.

> ## Pre-requisites

Reviewer 2: hmm that can work as well

`apify-client` requires Node.js version 16 or higher. Node.js is available for download on the [official website](https://nodejs.org/). Check for your current node version by running:

```bash
node -v
```

## Installation

You can install the client via [NPM](https://www.npmjs.com/) or use any other package manager of your choice.

```bash
npm i apify-client
# or
pnpm add apify-client
# or
yarn add apify-client
# or
bun add apify-client
```

## Authentication and Initialization

> To use the client, you need an [API token](https://docs.apify.com/platform/integrations/api#api-token). You can find your token in under integrations in [Apify Console](https://console.apify.com/account/integrations). Copy the token add initialize the client by providing the token (`MY-APIFY-TOKEN`) as a parameter to the `ApifyClient` constructor.

Reviewer: ```suggestion
To use the client, you need an [API token](https://docs.apify.com/platform/integrations/api#api-token). You can find your token under integrations in [Apify Console](https://console.apify.com/account/integrations). Copy the token and initialize the client by providing the token (`MY-APIFY-TOKEN`) as a parameter to the `ApifyClient` constructor.
```

> To use the client, you need an [API token](https://docs.apify.com/platform/integrations/api#api-token). You can find your token under integrations in [Apify Console](https://console.apify.com/account/integrations). Copy the token and initialize the client by providing the token (`MY-APIFY-TOKEN`) as a parameter to the `ApifyClient` constructor.

Reviewer 2: ```suggestion
To use the client, you need an [API token](https://docs.apify.com/platform/integrations/api#api-token). You can find your token under [Integrations](https://console.apify.com/account/integrations) tab in Apify Console. Copy the token and initialize the client by providing the token (`MY-APIFY-TOKEN`) as a parameter to the `ApifyClient` constructor.
```

```js
// ES5 example import
const { ApifyClient } = require('apify-client');
```

```js
// ES6+ example import
import { ApifyClient } from 'apify';
```

```js
// Client initialization with the API token
const client = new ApifyClient({
    token: 'MY-APIFY-TOKEN',
});
```

> > ❗ The API token is used to authorize your requests to Apify API. You can be charged for the usage of the underlying services so do not share your API token with untrusted parties or expose it on the client side of your applications

Reviewer: ```suggestion
> ❗ The API token is used to authorize your requests to the Apify API. You can be charged for the usage of the underlying services, so do not share your API token with untrusted parties or expose it on the client side of your applications
```

> > ❗ The API token is used to authorize your requests to the Apify API. You can be charged for the usage of the underlying services, so do not share your API token with untrusted parties or expose it on the client side of your applications

Reviewer 2: Admonitions are much better at catching readers eye than block quotes

```suggestion

:::warning Secure access

The API token is used to authorize your requests to the Apify API. You can be charged for the usage of the underlying services, so do not share your API token with untrusted parties or expose it on the client side of your applications

:::

```

## Quick start

> One of the most common use cases is starting [Actors](https://docs.apify.com/platform/actors) (serverless program running in the [Apify cloud](https://docs.apify.com/platform)) and getting results from their [datasets](https://docs.apify.com/platform/storage/dataset) (storage) after they finish the job (usually scraping, automation processes or data processing).

Reviewer: ```suggestion
One of the most common use cases is starting [Actors](https://docs.apify.com/platform/actors) (serverless programs running in the [Apify cloud](https://docs.apify.com/platform)) and getting results from their [datasets](https://docs.apify.com/platform/storage/dataset) (storage) after they finish the job (usually scraping, automation processes or data processing).
```

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

// Starts an Actor and waits for it to finish
const { defaultDatasetId } = await client.actor('username/actor-name').call();

// Lists items from the Actor's dataset
const { items } = await client.dataset(defaultDatasetId).listItems();
```

### Running Actors

> To start the Actor, you can use the [ActorClient](https://docs.apify.com/api/client/js/reference/class/ActorClient) (`client.actor()`) and pass the Actor's ID (e.g. `john-doe/my-cool-actor`) to define which Actor you want to run. The Actor's ID is a combination of the username and the Actor owner’s username. You can run both your own Actors and [Actors from Apify Store](https://docs.apify.com/platform/actors/running/actors-in-store).

Reviewer: ```suggestion
To start an Actor, you can use the [ActorClient](https://docs.apify.com/api/client/js/reference/class/ActorClient) (`client.actor()`) and pass the Actor's ID (e.g. `john-doe/my-cool-actor`) to define which Actor you want to run. The Actor's ID is a combination of the username and the Actor owner’s username. You can run both your own Actors and [Actors from Apify Store](https://docs.apify.com/platform/actors/running/actors-in-store).
```

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

// Runs an Actor with an input and waits for it to finish.
const { defaultDatasetId } = await client.actor('username/actor-name').call({
    some: 'input',
});
```

### Getting results from the dataset

To get the results from the dataset, you can use the [DatasetClient](https://docs.apify.com/api/client/js/reference/class/DatasetClient) (`client.dataset()`) and [`listItems()`](https://docs.apify.com/api/client/js/reference/class/DatasetClient#listItems) method. You need to pass the dataset ID to define which dataset you want to access. You can get the dataset ID from the Actor's run object (represented by `defaultDatasetId`).

```js

const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

// Lists items from the Actor's dataset.
const { items } = await client.dataset('dataset-id').listItems();
```

> > 💡 **Good to know**: Running an Actor might take time depending on the Actor's complexity and the amount of data it processes. If you want only to get data and have an immediate response you should access the existing dataset of the finished [Actor run](https://docs.apify.com/platform/actors/running/runs-and-builds#runs).

Reviewer: ```suggestion
> 💡 **Good to know**: Running an Actor might take time, depending on the Actor's complexity and the amount of data it processes. If you want only to get data and have an immediate response you should access the existing dataset of the finished [Actor run](https://docs.apify.com/platform/actors/running/runs-and-builds#runs).
```

> > 💡 **Good to know**: Running an Actor might take time, depending on the Actor's complexity and the amount of data it processes. If you want only to get data and have an immediate response you should access the existing dataset of the finished [Actor run](https://docs.apify.com/platform/actors/running/runs-and-builds#runs).

Reviewer 2: ```suggestion

:::note Dataset access

Running an Actor might take time, depending on the Actor's complexity and the amount of data it processes. If you want only to get data and have an immediate response you should access the existing dataset of the finished [Actor run](https://docs.apify.com/platform/actors/running/runs-and-builds#runs).

:::

```

## Usage concepts

The `ApifyClient` interface follows a generic pattern that applies to all of its components. By calling individual methods of `ApifyClient`, specific clients that target individual API resources are created. There are two types of those clients:

- [`actorClient`](https://docs.apify.com/api/client/js/reference/class/ActorClient): a client for the management of a single resource
- [`actorCollectionClient`](https://docs.apify.com/api/client/js/reference/class/ActorCollectionClient): a client for the collection of resources

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

// Collection clients do not require a parameter.
const actorCollectionClient = apifyClient.actors();
// Creates an actor with the name: my-actor.
const myActor = await actorCollectionClient.create({ name: 'my-actor-name' });
// List all your used Actors (both own and from Apify Store)
const { items } = await actorCollectionClient.list();
```

> > The resource ID can be either the `id` of the said resource, or a combination of your `username/resource-name`.

Reviewer 2: ```suggestion

:::note Resource identification

The resource ID can be either the `id` of the said resource, or a combination of your `username/resource-name`.

:::

```

```

The quick access to `dataset` and other storage directly from the run client can be used with the [`lastRun()`](https://docs.apify.com/api/client/js/reference/class/ActorClient#lastRun) method.

## Features

Based on the endpoint, the client automatically extracts the relevant data and returns it in the expected format. Date strings are automatically converted to `Date` objects. For exceptions, the client throws an [`ApifyApiError`](https://docs.apify.com/api/client/js/reference/class/ApifyApiError), which wraps the plain JSON errors returned by API and enriches them with other contexts for easier debugging.

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

try {
   const { items } = await client.dataset("non-existing-dataset-id").listItems();
} catch (error) {
   // The error is an instance of ApifyApiError
   const { message, type, statusCode, clientMethod, path } = error;
   // Log error for easier debugging 
   console.log({ message, statusCode, clientMethod, type });
}
```

### Retries with [exponential backoff](https://docs.apify.com/api/client/js/docs/features#retries-with-exponential-backoff)

> Network communication sometimes fails, that's a given. The client will automatically retry requests that failed due to a network error, an internal error of the Apify API (HTTP 500+), or a rate limit error (HTTP 429). By default, it will retry up to 8 times. The first retry will be attempted after ~500ms, the second after ~1000ms, and so on. You can configure those parameters using the `maxRetries` and `minDelayBetweenRetriesMillis` options of the `ApifyClient` constructor.

Reviewer: ```suggestion
Network communication sometimes fails. That's a given. The client will automatically retry requests that failed due to a network error, an internal error of the Apify API (HTTP 500+), or a rate limit error (HTTP 429). By default, it will retry up to 8 times. The first retry will be attempted after ~500ms, the second after ~1000ms, and so on. You can configure those parameters using the `maxRetries` and `minDelayBetweenRetriesMillis` options of the `ApifyClient` constructor.
```

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ 
  token: 'MY-APIFY-TOKEN',
  maxRetries: 8,
  minDelayBetweenRetriesMillis: 500, // 0.5s
  timeoutSecs: 360 // 6 mins
});
```

### Convenience functions and options

Some actions can't be performed by the API itself, such as indefinite waiting for an Actor run to finish (because of network timeouts). The client provides convenient `call()` and `waitForFinish()` functions that do that. If the limit is reached, the returned promise is resolved to a run object that will have status `READY` or `RUNNING` and it will not contain the Actor run output.

> [Key-value store](https://docs.apify.com/platform/storage/key-value-store) records can be retrieved as objects, buffers, or streams via the respective options, dataset items can be fetched as individual objects or serialized data and we plan to add better stream support and async iterators.

Reviewer 2: Please do not mention future plans in documentation.

```suggestion
[Key-value store](https://docs.apify.com/platform/storage/key-value-store) records can be retrieved as objects, buffers, or streams via the respective options, dataset items can be fetched as individual objects or serialized data.
```

Reviewer 3: That was part of the content we already have, I also wanted to remove it 👍

---
sidebar_label: 'Getting started'
title: 'Getting started'
---

# Apify API client for JavaScript

`apify-client` is the official library to access the [Apify REST API](https://docs.apify.com/api/v2) from your
JavaScript/TypeScript applications. It runs both in Node.js and browser and provides useful features like

> automatic retries and convenience functions that improve the experience of using the Apify API. All requests and responses (including errors) are encoded in JSON format with UTF-8 encoding.

Reviewer 4: nit: do not use prose wrap, let the editor handle line wraps when appropriate

## Pre-requisites

`apify-client` requires Node.js version 16 or higher. Node.js is available for download on the [official website](https://nodejs.org/). Check for your current node version by running:

```bash
node -v
```

## Installation

You can install the client via [NPM](https://www.npmjs.com/) or use any other package manager of your choice.

```bash
npm i apify-client
# or
pnpm add apify-client
# or
yarn add apify-client
# or
bun add apify-client
```

## Authentication and Initialization

To use the client, you need an [API token](https://docs.apify.com/platform/integrations/api#api-token). You can find your token under [Integrations](https://console.apify.com/account/integrations) tab in Apify Console. Copy the token and initialize the client by providing the token (`MY-APIFY-TOKEN`) as a parameter to the `ApifyClient` constructor.

```js
// ES5 example import
const { ApifyClient } = require('apify-client');
```

```js
// ES6+ example import
import { ApifyClient } from 'apify';

> ```

Reviewer 4: i wouldn't bother with CJS examples, lets use import/export everywhere, as we do in all the other places (and if we have some places where we still use require, we should change them too). if you absolutely want to mention the CJS support, do it once via admonition and keep all the examples with ESM imports, but i don't think its reasonable.

also its CJS (commonjs) and ESM (ecmascript modules), calling this ES5 and ES6 feels pretty confusing to me

Reviewer 3: Yeah, agree. We took inspiration from other docs. Let's unify it.

```js
// Client initialization with the API token
const client = new ApifyClient({
    token: 'MY-APIFY-TOKEN',
});
```

> 

Reviewer 4: nit: we should never have more than one blank line padding

i hope those things will be resolved automatically at some point (e.g. via biome or prettier)

Reviewer 3: It was introduced by applying suggestions from PR 😞 I reviewed all the extra spaces once more locally.

:::warning Secure access

The API token is used to authorize your requests to the Apify API. You can be charged for the usage of the underlying services, so do not share your API token with untrusted parties or expose it on the client side of your applications

:::

## Quick start

> One of the most common use cases is starting [Actors](https://docs.apify.com/platform/actors) (serverless programs running in the [Apify cloud](https://docs.apify.com/platform)) and getting results from their [datasets](https://docs.apify.com/platform/storage/dataset) (storage) after they finish the job (usually scraping, automation processes or data processing).

Reviewer 4: nit: links to the docs don't have to be absolute, e.g. `/platform/actors` will also work

Reviewer 3: Fixed all the absolute links

const runsClient = actorClient.runs();
// Lists the last 10 runs of your Actor.
const { items } = await runsClient.list({
    limit: 10,
    desc: true
});

// Select the last run of your Actor that finished
// with a SUCCEEDED status.
const lastSucceededRunClient = actorClient.lastRun({ status: 'SUCCEEDED' });
// Fetches items from the run's dataset.
const { items } = await lastSucceededRunClient.dataset()
    .listItems();
```

The quick access to `dataset` and other storage directly from the run client can be used with the [`lastRun()`](https://docs.apify.com/api/client/js/reference/class/ActorClient#lastRun) method.

## Features

Based on the endpoint, the client automatically extracts the relevant data and returns it in the expected format. Date strings are automatically converted to `Date` objects. For exceptions, the client throws an [`ApifyApiError`](https://docs.apify.com/api/client/js/reference/class/ApifyApiError), which wraps the plain JSON errors returned by API and enriches them with other contexts for easier debugging.

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

try {

>    const { items } = await client.dataset("non-existing-dataset-id").listItems();

Reviewer 4: bad indent (3 vs 4 spaces)

} catch (error) {
   // The error is an instance of ApifyApiError
   const { message, type, statusCode, clientMethod, path } = error;
   // Log error for easier debugging 
   console.log({ message, statusCode, clientMethod, type });
}
```

> ### Retries with [exponential backoff](https://docs.apify.com/api/client/js/docs/features#retries-with-exponential-backoff)

Reviewer 4: this link actually feels like it points to itself? the features page is now removed by this PR and the content you are linking is exactly here (unless i am missing something)

Network communication sometimes fails. That's a given. The client will automatically retry requests that failed due to a network error, an internal error of the Apify API (HTTP 500+), or a rate limit error (HTTP 429). By default, it will retry up to 8 times. The first retry will be attempted after ~500ms, the second after ~1000ms, and so on. You can configure those parameters using the `maxRetries` and `minDelayBetweenRetriesMillis` options of the `ApifyClient` constructor.

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ 

>   token: 'MY-APIFY-TOKEN',
>   maxRetries: 8,
>   minDelayBetweenRetriesMillis: 500, // 0.5s
>   timeoutSecs: 360 // 6 mins

Reviewer 4: bad indent
```suggestion
    token: 'MY-APIFY-TOKEN',
    maxRetries: 8,
    minDelayBetweenRetriesMillis: 500, // 0.5s
    timeoutSecs: 360 // 6 mins
```

});
```

### Convenience functions and options

Some actions can't be performed by the API itself, such as indefinite waiting for an Actor run to finish (because of network timeouts). The client provides convenient `call()` and `waitForFinish()` functions that do that. If the limit is reached, the returned promise is resolved to a run object that will have status `READY` or `RUNNING` and it will not contain the Actor run output.

[Key-value store](https://docs.apify.com/platform/storage/key-value-store) records can be retrieved as objects, buffers, or streams via the respective options, dataset items can be fetched as individual objects or serialized data.

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

// Starts an Actor and waits for it to finish.
const finishedActorRun = await client.actor('username/actor-name').call();

// Starts an Actor and waits maximum 60s for the finish 
const { status } = await client.actor('username/actor-name').start({

>   waitForFinish: 60, // 1 minute

Reviewer 4: bad indent
```suggestion
    waitForFinish: 60, // 1 minute
```

});
```

### Pagination

Most methods named `list` or `listSomething` return a [`Promise<PaginatedList>`](https://docs.apify.com/api/client/js/reference/interface/PaginatedList). There are some exceptions though, like `listKeys` or `listHead` which paginate differently. The results you're looking for are always stored under `items` and you can use the `limit` property to get only a subset of results. Other props are also available, depending on the method.

```js
const { ApifyClient } = require('apify-client');

const apifyClient = new ApifyClient({ token: 'MY-APIFY-TOKEN' });

// Resource clients accept an ID of the resource.
const datasetClient = apifyClient.dataset('dataset-id');

>  // Number of items per page

Reviewer 4: bad indent
```suggestion
// Number of items per page
```

const limit = 1000;
// Initial offset
let offset = 0; 
// Array to store all items
let allItems = []; 

while (true) {
    const { items, total } = await datasetClient.listItems({ limit, offset });
    
    console.log(`Fetched ${items.length} items`);
    
    // Merge new items with other already loaded items
    allItems.push(...items);
  

>   // If there are no more items to fetch, exit the loading

Reviewer 4: bad indent
```suggestion
    // If there are no more items to fetch, exit the loading
```

JavaScript applications. It runs both in Node.js and browser and provides useful features like
automatic retries and convenience functions that improve the experience of using the Apify API.

## Pre-requisites

`apify-client` requires Node.js version 16 or higher. Node.js is available for download on the [official website](https://nodejs.org/).

## Installation

> You can install the client via the [npm package](https://www.npmjs.com/package/apify-client). To do that, simply run `npm i apify-client`.

Reviewer 2: ```suggestion
You can install the client via the [npm package](https://www.npmjs.com/package/apify-client). To do that, simply run.

```bash
npm i apify-client
```
```

const myActorRun = await actorClient.start();
```

```js
// Resource clients accept an ID of the resource.
const datasetClient = apifyClient.dataset('john-doe/my-dataset');
// Appends items to the end of john-doe/my-dataset.
await datasetClient.pushItems([{ foo: 1 }, { bar: 2 }]);
```

> The ID of the resource can be either the `id` of the said resource, or a combination of your `username/resource-name`.

This is really all you need to remember, because all resource clients follow the pattern you see above.

## Nested clients

Sometimes clients return other clients. That's to simplify working with nested collections, such as runs of a given actor.

```js
const actorClient = apifyClient.actor('john-doe/hello-world');
const runsClient = actorClient.runs();
// Lists the last 10 runs of the john-doe/hello-world actor.
const { items } = await runsClient.list({
    limit: 10,
    desc: true
})

// Selects the last run of the john-doe/hello-world actor that finished
// with a SUCCEEDED status.
const lastSucceededRunClient = actorClient.lastRun({ status: 'SUCCEEDED' });
// Fetches items from the run's dataset.
const { items } = await lastSucceededRunClient.dataset()
    .listItems();
```

> > The quick access to `dataset` and other storages directly from the run client can now only be used with the `lastRun()` method, but the feature will be available to all runs in the future.

Reviewer 2: ```suggestion
:::note Storage quick access

The quick access to `dataset` and other storages directly from the run client can now only be used with the `lastRun()` method, but the feature will be available to all runs in the future.

:::
```
